File: native/engine/core/common/IDGen.cpp
#include "IDGen.h"

int IDGen::get()
{
    if (!pool.empty())
    {
        int id = pool[pool.size() - 1];
        pool.pop_back();
        return id;
    }
    return ++id;
}

void IDGen::recvoery(int id)
{
    if (pool.capacity() <= pool.size())
    {
        pool.reserve(pool.capacity() + 100);
    }
    pool.push_back(id);
}

File: sq/core/common/IDGen.ts

export default class IDGen{
   private pool:Array<number> = [];
   private id:number = 1;
   public get():number{
    if(this.pool.length > 0){
        return this.pool.pop();
    }
    return this.id++;
   }
   
   public recovery(id:number){
     this.pool.push(id);
   }
}